h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java

```diff
@@ -0,0 +1,153 @@
+
+import hex.tfidf.DocumentFrequency;
+import hex.tfidf.InverseDocumentFrequency;
+import hex.tfidf.TermFrequency;
+import hex.tfidf.TfIdfPreprocessor;
+import water.Key;
+import water.MRTask;
+import water.fvec.*;
+import water.rapids.Env;
+import water.rapids.Merge;
+import water.rapids.Val;
+import water.rapids.ast.AstPrimitive;
+import water.rapids.ast.AstRoot;
+import water.rapids.vals.ValFrame;
+
+/**
+ * Primitive AST operation to compute TF-IDF values for given document corpus.
+ */
+public class AstTfIdf extends AstPrimitive<AstTfIdf> {
+
+    /**
+     * Name to be used for a column containing Inverse Document Frequency values in the output frame of this operation.
+     */
+    private static final String IDF_COL_NAME = "IDF";
+    /**
+     * Name to be used for a column containing TF-IDF values in the output frame of this operation.
+     */
+    private static final String TF_IDF_COL_NAME = "TF-IDF";
+    /**
+     * Column names to be used for preprocessed frame.
+     */
+    private static final String[] PREPROCESSED_FRAME_COL_NAMES = new String[] { "DocID", "Word" };
+
+    @Override
+    public int nargs() {
+        return 1 + 2; // (tf-idf input_frame_name preprocess)
+    }
+
+    @Override
+    public String[] args() {
+        return new String[]{ "frame", "preprocess" };
+    }
+
+    @Override
+    public Val apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
+        Frame inputFrame = stk.track(asts[1].exec(env).getFrame());
+        boolean preprocess = asts[2].exec(env).getBool();
+
+        // Pre-processing
+        Frame wordFrame;
+        long documentsCnt;
+        if (preprocess) {
+            byte[] outputTypes = new byte[]{ Vec.T_NUM, Vec.T_STR };
+
+            TfIdfPreprocessor preprocessor = new TfIdfPreprocessor();
+            wordFrame = preprocessor.doAll(outputTypes, inputFrame).outputFrame(PREPROCESSED_FRAME_COL_NAMES, null);
+
+            documentsCnt = inputFrame.numRows();
+        } else {
+            wordFrame = inputFrame;
+            // TODO: Avoid converting to categorical Vec
+            documentsCnt = inputFrame.vec(0).toCategoricalVec().cardinality();
+        }
+
+        // DF
+        DocumentFrequency dfTaks = new DocumentFrequency();
+        Frame dfOutFrame = dfTaks.compute(wordFrame);
+
+        // IDF
+        InverseDocumentFrequency idf = new InverseDocumentFrequency(documentsCnt);
+        Vec idfValues = idf.doAll(new byte[]{ Vec.T_NUM }, dfOutFrame.lastVec()).outputFrame().anyVec();
+        // Replace DF column with IDF column
+        dfOutFrame.remove(dfOutFrame.numCols() - 1);
+        dfOutFrame.add(IDF_COL_NAME, idfValues);
+
+        // TF
+        TermFrequency tfTask = new TermFrequency();
+        Frame tfOutFrame = tfTask.compute(wordFrame);
+        
+        // Intermediate frame containing both TF and IDF values
+        // TODO: Avoid converting a string column to categorical
+        tfOutFrame.replace(1, tfOutFrame.vecs()[1].toCategoricalVec());
```

Comment:
could use AstUnique here as well, also I see potential leaks all over this method

Reply:
In this case I need to merge `tfOutFrame` and `dfOutFrame`, where term frequency values are "per word and document" and document frequency values which are "per word". Therefore, I need to add one value from `dfOutFrame` to multiple rows in `tfOutFrame`. How would you apply `AstUnique` to solve this?

I checked and it seems that long time ago there has been an implementation of `AstMerge` that worked with string values - see `mergetest.py` in https://github.com/h2oai/h2o-3/commit/f645ee8c6d86749845d033c4d1e4851bb9c96d52, but this no longer works.

But for now, it seems that the main reason why I was worried about conversion to `Categorical` vector - the `Categorical.MAX_CATEGORICAL_COUNT` does not apply for `String` vectors that are being transformed to `Categorical` vector. I tried to convert `String` vector containing `Categorical.MAX_CATEGORICAL_COUNT + 10` unique values to `Categorical` vector and it worked just fine.

Is there a better solution for this?

Also, I fixed the leaks by using `Scope`.

Reply:
Converting to categorical and doing merge is just hiding the problem. The problem is that you need to do a merge of two frames over a very large string-based key.
From my quick read of the code both tfOutFrame and dfOutFrame are computed from the same source wordsFrame, it might be possible to make this more efficient by combining both of the computations of tfOut and dfOut into a single series of MRTasks.
On the other hand how big can the domain really be? We are taking about text here, so the domains should not be bigger than couple 100k. So maybe the categorical+merge is not such a big deal.
@michalkurka any thoughts?

Reply:
I know. Well, yes that is true, both `tfOutFrame` and `dfOutFrame` are built by using same input, but it is done via `GroupBy`, where `TermFrequency` is done on grouping by `docID` and `word`, and `DocumentFrequency` first groups by the same columns (I changed it in the last commit to expecte unique words per document, therefore reuse the output of `TermFrequencyTask` since it is not a publicly available API and it saves some computation.) and then do another groupby operation in order to count number of documents containing given word.

Well, to be honest it looks like that even `Categorical.MAX_CATEGORICAL_COUNT` shouldn't be a problem even for larger document corpora. But if there is a better way to solve this I would prefer that over converting it to a categorical vector.